src/freenet/node/NewPacketFormat.java

```diff
@@ -590,71 +590,28 @@ NPFPacket createPacket(int maxPacketSize, PeerMessageQueue messageQueue, Session
 			if(logDEBUG) Logger.debug(this, "Added acks for "+this+" for "+pn.shortToString());
 		}
 		
-		byte[] haveAddedStatsBulk = null;
-		byte[] haveAddedStatsRT = null;
-		
 		if(!ackOnly) {
 			
 			boolean addedFragments = false;
 			
-			while(true) {
-				
-				boolean addStatsBulk = false;
-				boolean addStatsRT = false;
-				
-				synchronized(sendBufferLock) {
-					// Always finish what we have started before considering sending more packets.
-					// Anything beyond this is beyond the scope of NPF and is PeerMessageQueue's job.
-addOldLoop:			for(Map<Integer, MessageWrapper> started : startedByPrio) {
-						//Try to finish messages that have been started
-						Iterator<MessageWrapper> it = started.values().iterator();
-						while(it.hasNext() && packet.getLength() < maxPacketSize) {
-							MessageWrapper wrapper = it.next();
-							while(packet.getLength() < maxPacketSize) {
-								MessageFragment frag = wrapper.getMessageFragment(maxPacketSize - packet.getLength());
-								if(frag == null) break;
-								mustSend = true;
-								addedFragments = true;
-								packet.addMessageFragment(frag);
-								sentPacket.addFragment(frag);
-								if(wrapper.allSent()) {
-									if((haveAddedStatsBulk == null) && wrapper.getItem().sendLoadBulk) {
-										addStatsBulk = true;
-										// Add the lossy message outside the lock.
-										break addOldLoop;
-									}
-									if((haveAddedStatsRT == null) && wrapper.getItem().sendLoadRT) {
-										addStatsRT = true;
-										// Add the lossy message outside the lock.
-										break addOldLoop;
-									}
-								}
-							}
+			synchronized(sendBufferLock) {
+				// Always finish what we have started before considering sending more packets.
+				// Anything beyond this is beyond the scope of NPF and is PeerMessageQueue's job.
```

Comment:
This may cause changes in packet sending, so it could be critical. Do you see any code path where this could change the effective ordering/prioritization of messages (in a bad way)?

Reply:
It's a lot of code touched, I see that. Not sure if that can be split up in smaller chunks.

There should not be any effective change, as all of these code paths are effectively unused. In this file specifically, the code paths for adding lossy messages are removed (those were used exclusively for load stats). Unfortunately, that code was quite convoluted.

I effectively started by removing this method in PeerNode:
```java
	@Override
	public MessageItem makeLoadStats(boolean realtime, boolean boostPriority, boolean noRemember) {
	    // FIXME re-enable when try NLM again.
	    return null;
```
And then pruned my way down all code that became effectively unreachable or effectively no-op.
* assume all invocations of `pn.makeLoadStats(...)` return `null` (which is clearly true as per the above FIXME)
* consider all instances of `sendStatsBulk`, `sendStatsRT` and their variations to be effectively `false`, as there is nothing to send
* consider all variations that could register the need to send stats (like `setSendLoadStatsASAP(...)`) no-ops
* every check that solely guards the above can now be removed

Then the changes probably make more sense :smile: